### DISCUSSION TITLE
inference on whole video but still show most frames unpredicted / without instances

### DISCUSSION THREAD

**Discussion by Cong-SUN** [talmolab/sleap#1359] [Help!]
Hey I run inference on the whole video and it worked (popping window says all frames are labeled with instance), but then in the GUI, it still says that only 0.01% of the frames are predicted. At the meantime, I already have a whole xxxxx.predictions.slp file in my predictions folder. So I'm confused as to whether all the frames in the video have a predicted instance or not.

**Comment by roomrys** [comment]
Hi @Cong-SUN,

A few questions:

1. Can you verify that there are predicted instances on the video you ran inference on (by quickly scrolling through the frames in the video of interest)? 
2. Also, can you provide a screenshot of your GUI (including the seekbar and the bottom display)? 
3. Is this just a problem with the calculation of the percentage, or are you not seeing any predictions on that video? 
4. How do the prediction look like in the predictions.slp file?

Thanks,
Liezl

**Reply by Cong-SUN** [reply]
Thanks for the quick reply, Liezl!
1. yes there are predicted instances on the video, but I don't think that's from the "run inference on whole video" trial. Instead, it seems to be from the previous "run inference on random 20 frames from this video" trial
2. ![248036675-a7339016-9c8f-486d-ae3a-88a9440e14ba](https://github.com/talmolab/sleap/assets/38435167/f81e26a5-0b17-4f6b-ba3e-2927049493cb)

3. the same as question 1
4. I inspected the predictions.slp file using `sleap-inspect` and it returns this:
```shell
Labeled frames: 299998
Tracks: 0
Video files:
  C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230403_CA1103195_Patch(0)-FireWire-0.mp4
    labeled frames: 299998
    labeled frames from 0 to 299997
    user labeled frames: 0
    tracks: 1
    max instances in frame: 1
Total user labeled frames: 0

Provenance:
  model_paths: ['C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\\models\\test_11_230621_145849.single_instance.n=260\\training_config.json']
  predictor: SingleInstancePredictor
  sleap_version: 1.3.0
  platform: Windows-10-10.0.17763-SP0
  command: C:\casa_phy\Miniconda3\envs\sleap\Scripts\sleap-track C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/labels.v001.slp --video.index 6 --frames 0,-299997 -m C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\models\test_11_230621_145849.single_instance.n=260\training_config.json --tracking.tracker none -o C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\predictions\labels.v001.slp.230621_160526.predictions.slp --verbosity json --no-empty-frames
  data_path: C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/labels.v001.slp
  output_path: C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\predictions\labels.v001.slp.230621_160526.predictions.slp
  total_elapsed: 5040.254548072815
  start_timestamp: 2023-06-21 16:05:35.382140
  finish_timestamp: 2023-06-21 17:29:35.636688
  args: {'data_path': 'C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/labels.v001.slp', 'models': ['C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\\models\\test_11_230621_145849.single_instance.n=260\\training_config.json'], 'frames': '0,-299997', 'only_labeled_frames': False, 'only_suggested_frames': False, 'output': 'C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun\\predictions\\labels.v001.slp.230621_160526.predictions.slp', 'no_empty_frames': True, 'verbosity': 'json', 'video.dataset': None, 'video.input_format': 'channels_last', 'video.index': '6', 'cpu': False, 'first_gpu': False, 'last_gpu': False, 'gpu': 'auto', 'max_edge_length_ratio': 0.25, 'dist_penalty_weight': 1.0, 'batch_size': 4, 'open_in_gui': False, 'peak_threshold': 0.2, 'tracking.tracker': 'none', 'tracking.target_instance_count': None, 'tracking.pre_cull_to_target': None, 'tracking.pre_cull_iou_threshold': None, 'tracking.post_connect_single_breaks': None, 'tracking.clean_instance_count': None, 'tracking.clean_iou_threshold': None, 'tracking.similarity': None, 'tracking.match': None, 'tracking.robust': None, 'tracking.track_window': None, 'tracking.min_new_track_points': None, 'tracking.min_match_points': None, 'tracking.img_scale': None, 'tracking.of_window_size': None, 'tracking.of_max_levels': None, 'tracking.save_shifted_instances': None, 'tracking.kf_node_indices': None, 'tracking.kf_init_frame_count': None}
``` 

**importantly: I just tried running inference on the whole video for a much shorter video, and it turns out to be just fine! --  Every frame is shown to be predicted.**
The videos that went wrong all have 300,000 frames -- do you think that would be the problem?

**Reply by roomrys** [reply]
Hi @Cong-SUN,

Inference via the GUI works in two stages, first prediction are created from the model and saved to the predictions.slp, then the predictions.slp is merged in with the original slp loaded in the GUI.

It is strange that the predictions were merged in for some videos but not others... Sorry, more questions since I am still scratching my head:
1. Are these videos (the long videos that had trouble merging and the shorter ones that were fine) in the same project? 
2. After failure merging the predictions automatically post-inference, did anything pop up in the terminal?
3. Are you able to successfully merge the predictions in manually via File > Merge Into Project

Thanks,
Liezl

**Reply by Cong-SUN** [reply]
Hi Liezl,

Thanks for letting me know. I didn't realize I need to merge the prediction into current project to visualize them in the GUI!

1. Yes those videos are in the same project. (but added in different times)
2. I didn't notice anything pop up in the terminal, but I'm not sure I checked closely. I am trying another long video and we'll see after that.
3. I think it successfully merged, but the GUI is a bit weird: it automatically added another short pseudo-video when merging the long video --- I merged the no.5 video, but it adds a no.7 for me... Then it shows for this new short pseudo-video which is only pseudoly 1268 frames long that 299998 frames have been predicted... and if you go to no.5 video which is the long video that's actually predicted and merged, it still shows that most of the frames are not predicted.
![image](https://github.com/talmolab/sleap/assets/122485661/b2b8bf8f-d659-44a8-88d6-35afef37dd28)
![image](https://github.com/talmolab/sleap/assets/122485661/0c25a7c1-922a-4ca7-b8bf-d3b6b6f2a991)

**Comment by roomrys** [comment]
I think we might be entering uncharted territory here. First thing to do would be to go back to the last saved version.

After checking out the command in [your other post](https://github.com/talmolab/sleap/discussions/1360):
> Heyy, I used this 
>````
>(sleap) C:\Windows\system32>
>sleap-track 
>"C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4" 
>--frames 0-1494 
>-m "C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/models/test_11_230621_145849.single_instance.n=260/best_model.h5" 
>[other_args] 
>--peak_threshold 0.05
>```` 
>to run inference but it gives this UnicodeDecodeError:

and 
>I didn't realize I need to merge the prediction into current project to visualize them in the GUI!

I think I know what might be happening. 

Are you running cli commands to do the inference instead of running inference through the GUI (and using Predict > Run Inference)? If you are running CLI commands to do the inference and would prefer to do it that way, then in order to automatically merge in the predictions (hopefully correctly, it looks like we got in a bit of a weird state above...) you need to use the path to the slp as the data path, not the path to the video. When using the slp as the data path, you can use the [`--video.index` argument](https://sleap.ai/guides/cli.html?highlight=cli#sleap-track) to specify which video you would like to run inference on.

Let me know if that was the case and how it goes.

Thanks,
Liezl

**Reply by Cong-SUN** [reply]
Ah no sorry... I was using GUI to run inference when encountering this "not automatically merging" issue; and was also using GUI to merge the predictions when encountering this "weirdly adding another pseudo short video after merging" issue.  
The CLI command was for rerunning inference with different peak_threshold setting for testing some further data analysis issue.

Sorry for the confusion!

Thank you,
Cong
